tests/test_trie.c

```diff
@@ -84,4 +84,173 @@ Test(trie, free)
 
     cr_assert_eq(rc, 0, "trie_free failed");
 
+}
+
+/* Checks that searching an empty trie returns 0 */
+Test(trie, empty_search)
+{
+    trie_t *t;
+    int found;
+    
+    t = new_trie('\0');
+    cr_assert_not_null(t, "new_trie() failed");
+
+    found = trie_search("b", t);
+
+    cr_assert_eq(found, 0, "trie_search() does not return 0 with an empty trie");
+}
+
+/* Checks a trie with only one word to see if trie_search returns expected for given str*/
+void singleton_search(char* str, int expected)
+{
+    trie_t *t;
+    int inserted, found;
+
+    t = new_trie('\0');
+    cr_assert_not_null(t, "new_trie() failed");
+
+    inserted = insert_string("CS220", t);
+    cr_assert_eq(inserted, 0, "insert_string() failed");
+
+    found = trie_search(str, t); // Checks if str can be found 
+    cr_assert_eq(found, expected, "trie_search() returns %s instead of %s with inserted word",
+            (found==0)? "NOT_IN_TRIE" : ((found==1)? "IN_TRIE" : "PARTIAL_IN_TRIE"),
+            (expected==0)? "NOT_IN_TRIE" : ((expected==1)? "IN_TRIE" : "PARTIAL_IN_TRIE"));
+}
+
+
+
+/* Checks for the inserted word in a singleton trie */
+Test(trie, singleton_search_inserted)
+{
+    singleton_search("CS220", IN_TRIE);
+}
+
+/* Checks for the prefix of the inserted word in a singleton trie */
+Test(trie, singleton_search_prefix)
+{
+    singleton_search("CS", PARTIAL_IN_TRIE);
+}
+
+
+/* Checks for the extension of the inserted word in a singleton trie */
+Test(trie, singleton_search_extension)
+{
+    singleton_search("CS22000", NOT_IN_TRIE);
+}
+
+/* Checks for a non-inserted word in a singleton trie */
+Test(trie, singleton_search_not_inserted)
+{
+    singleton_search("Computer", NOT_IN_TRIE);
+}
+
+/* Checks a trie with 21 words to see if trie_search returns expected for given str*/
+void twenty_search(char* str, int expected)
+{
+    trie_t *t;
+    int inserted, found;
+    char* words[21] = {"Ever", "loved", "someone", "so", "much,", "you", "would", "do", "anything", "for", "them?", "Yeah,", "well,", "make", "that", "yourself", "and", "whatever", "the", "hell", "want."}; // quote by Harvey Specter 
```

Comment:
So I hadn't noticed this until here, but you probably don't want to be manually stripping out duplicate words before calling insert. The insert function should silently succeed when given a word already in the trie, and this functionality should be tested. When teams go to insert huge quantities of text into the trie, they shouldn't have to ensure that they aren't calling insert twice on the same word (and in fact it won't even be possible: when the application teams start to buffer text, they won't know the next or previous lines).